[GStreamer][MSE] Support playing videos of less than 1 sec with libav decoders

eocanha opened the pull request:
#### bc7ce86c849cbc01c75e33c073e30e185b3455c8
<pre>
[GStreamer][MSE] Support playing videos of less than 1 sec with libav decoders
<a href="https://bugs.webkit.org/show_bug.cgi?id=293239">https://bugs.webkit.org/show_bug.cgi?id=293239</a>

Reviewed by NOBODY (OOPS!).

GStreamer video decoders based on libav are currently implemented in a
way that hoard one video frame per decoding thread. This not only means
more latency during continuous playback, but also on stalls (when no
more frames are supplied at the video decoder input). While problem has
been mitigated by reducing the max-threads to 2 [1], the behaviour still
causes problems in some tests with very short videos.

This commit checks that we&apos;re currently using the libav video decoder
and, if the video position hasn&apos;t moved but in theory should, emits a
still frame event to force the video decoder to emit all the video
frames at its source pad.

A layout test to check that the video decoder has emitted all the frames
when a stall happens has been added. This test explicitly doesn&apos;t check
the exact played timestamps, to not fail because of missing edit list
support, which skews the timestamps.

See also:
<a href="https://bugs.webkit.org/show_bug.cgi?id=270622">https://bugs.webkit.org/show_bug.cgi?id=270622</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=231019">https://bugs.webkit.org/show_bug.cgi?id=231019</a>

[1] <a href="https://github.com/WebKit/WebKit/commit/128e962baa61f49c78087a622eaecd08dc692ef6#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR2831">https://github.com/WebKit/WebKit/commit/128e962baa61f49c78087a622eaecd08dc692ef6#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR2831</a>

* LayoutTests/media/media-source/media-source-shortmediastall-expected.txt: Added.
* LayoutTests/media/media-source/media-source-shortmediastall.html: Added.
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h: Added the LibAv video decoding platform.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::playbackPosition const): Emit the still frame event if the video hasn&apos;t advanced but in theory should have, but only when using a libav decoder and when the player isn&apos;t stalled in the same timestamp as the last time (avoid redundant emission).
(WebCore::MediaPlayerPrivateGStreamer::configureVideoDecoder): Declare that we&apos;re using the LibAv video decoding platform when present.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h: Added m_lastStalledPosition to store the timestamp where the player got last stalled. This is used to avoid redundant emission of the still frame event on consecutive stalls that are in fact for the the same timestamp.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::hasFutureTime const): Expose MediaSourcePrivate::hasFutureTime() through the player private.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h: Add hasFutureTime().
* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(webKitMediaSrcSendEvent): Forward the GstEventStillFrame custom downstream event to the first (and only, if present) video src pad, but enqueing it in the buffer/event queue in MediaSourceTrackGStreamer.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc7ce86c849cbc01c75e33c073e30e185b3455c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111645 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31312 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21786 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117680 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61927 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113607 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32000 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39897 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84823 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35571 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114592 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25544 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100482 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65264 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24880 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18617 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61512 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94929 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18685 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120928 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38698 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28751 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93722 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39076 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96739 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93548 "Found 2 new API test failures: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy, /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38688 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16481 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34734 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38587 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44072 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38250 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41586 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39952 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/898db7db5b430d8739f68e9013adb4b79010276f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45600

eocanha commented:
Ok, but I'll have to be lax with the timestamp checks, because this patch/bug is a dependency of https://bugs.webkit.org/show_bug.cgi?id=270622, and that one depends on https://bugs.webkit.org/show_bug.cgi?id=231019 "[MSE][GStreamer] Honor MP4 edit lists". If it wasn't because of the (unsupported) edit lists timestamp skew, this commit would already have been the fix for 270622 "[GStreamer][MSE] playback doesn't start if less than 1s of video is available."

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
```diff
@@ -856,6 +856,22 @@ static gboolean webKitMediaSrcSendEvent(GstElement* element, GstEvent* eventTran
+            GST_DEBUG_OBJECT(element, "No video stream to send the still frame event to");
+            return FALSE;
+        }
+        FALLTHROUGH;
```
[[fallthrough]]

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1598,9 +1598,21 @@ MediaTime MediaPlayerPrivateGStreamer::playbackPosition() const
-    else if (m_canFallBackToLastFinishedSeekPosition)
+        if (isMediaSource() && m_videoDecoderPlatform == GstVideoDecoderPlatform::LibAv && playbackPosition == m_cachedPosition) {
+            auto* playerPrivateMSE = reinterpret_cast<MediaPlayerPrivateGStreamerMSE*>(const_cast<MediaPlayerPrivateGStreamer*>(this));
+            bool isPaused = paused();
+            bool hasFutureTime = playerPrivateMSE->hasFutureTime(m_cachedPosition);
```
no need for these 2 local vars

eocanha replied:
They made the code clearer in [MediaPlayerPrivateGStreamerMSE::timeIsProgressing()](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp#L585) the original place where the idea of that code comes from.
Unfortunately, we can't call `timeIsProgressing()` from `currentTime()` because that would cause infinite recursion, so we have to "reimplement" `timeIsProgressing()` inside `playbackPosition()` (which is the inner implementation of `currentTime()`).

But ok, I'll remove them.

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
```diff
@@ -856,6 +856,22 @@ static gboolean webKitMediaSrcSendEvent(GstElement* element, GstEvent* eventTran
         return forwardToAllPads ? wasEventHandledByAllStreams : wasEventHandledByAnyStream;
     }
+    case GST_EVENT_CUSTOM_DOWNSTREAM: {
+        if (gst_structure_has_name(gst_event_get_structure(event.get()), "GstEventStillFrame")) {
```
gst_event_has_name()

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
```diff
@@ -856,6 +856,22 @@ static gboolean webKitMediaSrcSendEvent(GstElement* element, GstEvent* eventTran
+                if (stream->track.get().type() == TrackPrivateBaseGStreamer::TrackType::Video) {
+                    GRefPtr<GstPad> sinkPeerPad = adoptGRef(gst_pad_get_peer(stream->pad.get()));
+                    stream->track->enqueueObject(adoptGRef(GST_MINI_OBJECT(event.leakRef())));
+                    return TRUE;
```
I guess it's unlikely, but what if we have multiple video tracks? only the first one is handled.

eocanha replied:
If we have more than one video track... we're in biiiiig trouble. That scenario hasn't ever been exercised in practical terms, even in the spec allows it in theory.

But ok, I'll change the code to continue looping and process all video tracks, just in case.

ntrrgc reviewed (commented):
I'm very happy to see a working workaround for this tricky issue, but I have several concerns with handling this inside `playbackPosition()`:

* What if `playbackPosition()` never gets called? The workaround would not be applied then.
  * In practical terms: does this work if the JS code is not polling `.currentTime`?
* What prevents this from unreasonably spamming the decoder with events?
  * The only rate limiting is linked to the currentTime cache, whose life expectancy is one main loop cycle.
  * In the worst case (we think there are buffers to drain, but draining is not advancing the time as we expect, and JS is polling `currentTime` once per loop) we would be sending a event every tick.

eocanha commented:
> * What if `playbackPosition()` never gets called? The workaround would not be applied then.

It's always called, because we can have two scenarios:

a) The player is paused by the user. In that case, who cares about decoder absolute latency? it just got paused wherever it was before. No real problem, it won't advance more, we don't need the extra frames.
b) The player was playing but got paused because of starvation. This is the case we want to solve. As the player was playing, we have periodic `playbackPosition()` checks that would trigger the detection (see my explanation on the next point). On top of that, there are also periodic `MediaSource::monitorSourceBuffers()` calls, with this call sequence:

- MediaSource::monitorSourceBuffers()  would detect HaveCurrentData or even "just" HaveFutureData in a very degenerated case of frame hoarding. This method *already* calls MediaSource --> MediaSourcePrivate::currentTime() --> MediaPlayerPrivateInterface::currentOrPendingSeekTime() --> MediaPlayerPrivateInterface::currentTime() --> MediaPlayerPrivateGStreamer::currentTime().
- MediaSourcePrivate::setMediaPlayerReadyState(), implemented by MediaSourcePrivateGStreamer::setMediaPlayerReadyState()
- MediaPlayerPrivateGStreamerMSE::setReadyState()
- MediaPlayerPrivateGStreamerMSE::propagateReadyStateToPlayer()
- Checks if `currentTime() == duration()`, so calls currentTime() too.

...but they aren't as important as the periodic time updates of a player where time is marching on.

> * In practical terms: does this work if the JS code is not polling `.currentTime`?

Yes. It's in HTMLMediaElement's best interest to poll currentTime() by itself. That's done by HTMLMediaElement::m_playbackProgressTimer, which triggers playbackProgressTimerFired(), which calls currentMediaTime(), which calls MediaPlayer::currentTime(), which calls MediaPlayerPrivateGStreamer::currentTime(), which calls playbackPosition().

> * What prevents this from unreasonably spamming the decoder with events?

Nothing, but my experiments show that the spamming is not so much. Setting a wait of 3 seconds after the first batch of appends, I can see two legitimate events sent and 5 spammy ones:

```
0:00:02.383403065 141426 0x560798d3efd0 DEBUG      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1611:playbackPosition:<MSE-media-player-0> The player seems stuck at {1749999999/1000000000 = 1.749999999}. Sent still frame event to force emission of the last decoded frames
0:00:05.637306672 141426 0x560798d3efd0 DEBUG      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1611:playbackPosition:<MSE-media-player-0> The player seems stuck at {2004729663/1000000000 = 2.004729663}. Sent still frame event to force emission of the last decoded frames
0:00:05.637529580 141426 0x560798d3efd0 DEBUG      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1611:playbackPosition:<MSE-media-player-0> The player seems stuck at {2004729663/1000000000 = 2.004729663}. Sent still frame event to force emission of the last decoded frames
0:00:05.638209464 141426 0x560798d3efd0 DEBUG      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1611:playbackPosition:<MSE-media-player-0> The player seems stuck at {2004729663/1000000000 = 2.004729663}. Sent still frame event to force emission of the last decoded frames
0:00:05.638255963 141426 0x560798d3efd0 DEBUG      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1611:playbackPosition:<MSE-media-player-0> The player seems stuck at {2004729663/1000000000 = 2.004729663}. Sent still frame event to force emission of the last decoded frames
0:00:05.638351991 141426 0x560798d3efd0 DEBUG      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1611:playbackPosition:<MSE-media-player-0> The player seems stuck at {2004729663/1000000000 = 2.004729663}. Sent still frame event to force emission of the last decoded frames
0:00:05.638514900 141426 0x560798d3efd0 DEBUG      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1611:playbackPosition:<MSE-media-player-0> The player seems stuck at {2004729663/1000000000 = 2.004729663}. Sent still frame event to force emission of the last decoded frames
```

Is it actually worth to increase the complexity of the solution and potentially create more corner cases to get rid of this inefficiency?

>   * The only rate limiting is linked to the currentTime cache, whose life expectancy is one main loop cycle.

That's great, because then we can re-evaluate the stall condition often.

>   * In the worst case (we think there are buffers to drain, but draining is not advancing the time as we expect, and JS is polling `currentTime` once per loop) we would be sending a event every tick.

Only for some extra ticks, according to the logs.

eocanha committed:
[GStreamer][MSE] Support playing videos of less than 1 sec with libav decoders
https://bugs.webkit.org/show_bug.cgi?id=293239

Reviewed by NOBODY (OOPS!).

GStreamer video decoders based on libav are currently implemented in a
way that hoard one video frame per decoding thread. This not only means
more latency during continuous playback, but also on stalls (when no
more frames are supplied at the video decoder input). While problem has
been mitigated by reducing the max-threads to 2 [1], the behaviour still
causes problems in some tests with very short videos.

This commit checks that we're currently using the libav video decoder
and, if the video position hasn't moved but in theory should, emits a
still frame event to force the video decoder to emit all the video
frames at its source pad.

A layout test to check that the video decoder has emitted all the frames
when a stall happens has been added. This test explicitly doesn't check
the exact played timestamps, to not fail because of missing edit list
support, which skews the timestamps.

See also:
https://bugs.webkit.org/show_bug.cgi?id=270622
https://bugs.webkit.org/show_bug.cgi?id=231019

[1] https://github.com/WebKit/WebKit/commit/128e962baa61f49c78087a622eaecd08dc692ef6#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR2831

* LayoutTests/media/media-source/media-source-shortmediastall-expected.txt: Added.
* LayoutTests/media/media-source/media-source-shortmediastall.html: Added.
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h: Added the LibAv video decoding platform.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::playbackPosition const): Emit the still frame event if the video hasn't advanced but in theory should have, but only when using a libav decoder and when the player isn't stalled in the same timestamp as the last time (avoid redundant emission).
(WebCore::MediaPlayerPrivateGStreamer::configureVideoDecoder): Declare that we're using the LibAv video decoding platform when present.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h: Added m_lastStalledPosition to store the timestamp where the player got last stalled. This is used to avoid redundant emission of the still frame event on consecutive stalls that are in fact for the the same timestamp.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::hasFutureTime const): Expose MediaSourcePrivate::hasFutureTime() through the player private.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h: Add hasFutureTime().
* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(webKitMediaSrcSendEvent): Forward the GstEventStillFrame custom downstream event to the first (and only, if present) video src pad, but enqueing it in the buffer/event queue in MediaSourceTrackGStreamer.

eocanha commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1598,9 +1598,23 @@ MediaTime MediaPlayerPrivateGStreamer::playbackPosition() const
         playbackPosition = MediaTime(gstreamerPosition, GST_SECOND);
-    else if (m_canFallBackToLastFinishedSeekPosition)
+        if (isMediaSource() && m_videoDecoderPlatform == GstVideoDecoderPlatform::LibAv) {
+            if (playbackPosition == m_cachedPosition && m_lastStalledPosition.isInvalid()) {
```
I did something even better than the test you proposed: I used `setInterval(function(){ ... video.currentTime; }, 0)`. The result is that currentTime is different every time and no stall is spuriously triggered. You can see that fine granularity in this filtered log fragment:

```
0:00:02.230341839 507241 0x55feb3e28d50 TRACE      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<MSE-v-0> Position 0:00:00.000000000, canFallBackToLastFinishedSeekPosition: false
0:00:02.231302654 507241 0x55feb3e28d50 TRACE      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<MSE-v-0> Position 0:00:00.000682518, canFallBackToLastFinishedSeekPosition: false
0:00:02.233202525 507241 0x55feb3e28d50 TRACE      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<MSE-v-0> Position 0:00:00.002582453, canFallBackToLastFinishedSeekPosition: false
0:00:02.235393967 507241 0x55feb3e28d50 TRACE      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<MSE-v-0> Position 0:00:00.004774475, canFallBackToLastFinishedSeekPosition: false
0:00:02.240210612 507241 0x55feb3e28d50 TRACE      webkitmediaplayer MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<MSE-v-0> Position 0:00:00.009589222, canFallBackToLastFinishedSeekPosition: false
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bc7ce86c849cbc01c75e33c073e30e185b3455c8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45600
